src/utils/random.rs

```diff
@@ -0,0 +1,266 @@
+/// The gap-sampling compute logic requires taking log(x), where x is sampled from an RNG.
+/// To guard against errors from taking log(0), a positive epsilon lower bound is applied.
+/// A good value for this parameter is at or near the minimum positive floating
+/// point value returned by for the RNG being used.
+// TODO: this is a straight port, it may not apply exactly to pcg64 rng but should be mostly fine;
+// double check; Apache Spark(tm) uses XORShift by default
+const RNG_EPSILON: f64 = 5e-11;
+
+/// Sampling fraction arguments may be results of computation, and subject to floating
+/// point jitter.  I check the arguments with this epsilon slop factor to prevent spurious
+/// warnings for cases such as summing some numbers to get a sampling fraction of 1.000000001
+const ROUNDING_EPSILON: f64 = 1e-6;
+
+type RSamplerFunc<'a, T> = Box<dyn Fn(Box<dyn Iterator<Item = T>>) -> Vec<T> + 'a>;
+
+pub(crate) trait RandomSampler<T: Data>: Send + Sync + Serialize + Deserialize {
+    /// Returns a function which returns random samples,
+    /// the sampler is thread-safe as the RNG is seeded with random seeds per thread.
+    fn get_sampler(&self) -> RSamplerFunc<T>;
+}
+
+pub(crate) fn get_default_rng() -> Pcg64 {
+    Pcg64::new(
+        0xcafe_f00d_d15e_a5e5,
+        0x0a02_bdbf_7bb3_c0a7_ac28_fa16_a64a_bf96,
+    )
+}
+
+/// Get a new rng with random thread local random seed
+fn get_rng_with_random_seed() -> Pcg64 {
+    Pcg64::seed_from_u64(rand::random::<u64>())
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub(crate) struct PoissonSampler {
+    fraction: f64,
+    use_gap_sampling_if_possible: bool,
+    prob: f64,
+}
+
+impl PoissonSampler {
+    pub fn new(fraction: f64, use_gap_sampling_if_possible: bool) -> PoissonSampler {
+        let prob = if fraction > 0.0 { fraction } else { 1.0 };
+
+        PoissonSampler {
+            fraction,
+            use_gap_sampling_if_possible,
+            prob,
+        }
+    }
+}
+
+impl<T: Data> RandomSampler<T> for PoissonSampler {
+    fn get_sampler(&self) -> RSamplerFunc<T> {
+        Box::new(move |items: Box<dyn Iterator<Item = T>>| -> Vec<T> {
+            if self.fraction <= 0.0 {
+                vec![]
+            } else {
+                let use_gap_sampling = self.use_gap_sampling_if_possible
+                    && self.fraction <= DEFAULT_MAX_GAP_SAMPLING_FRACTION;
+
+                let mut gap_sampling = if use_gap_sampling {
+                    // Initialize here and move to avoid constructing a new one each iteration
+                    Some(GapSamplingReplacement::new(self.fraction, RNG_EPSILON))
+                } else {
+                    None
+                };
+                let dist = Poisson::new(self.prob).unwrap();
+                let mut rng = get_rng_with_random_seed();
+
+                items
+                    .filter(move |item| {
+                        let count = if use_gap_sampling {
```

Comment:
This bit misses out the replication logic. So, basically for with replication case, the replication factor is drawn from the Poisson distribution.  You have handled the sampling gaps here(filtering out count == 0), but you didn't replicate when it is > 1. You can test this in a test case like the following. You have used the sampling size of 3 in tests for with replacement case, for which it will most likely provide enough samples. Try increasing the sample size to 9, and run it a few times, you will start seeing your warning take_sample method(Needed to re-sample due to insufficient sample size) in log messages. In case of a very large array with sample size near array.len(), you will see this going to a very long loop.

Reply:
Great catch. Yes this was an unfinished placeholder I forgot to update... I have updated the test to trigger drawing samples > 1 and added a guard against high repetition + fixed the logic.

Something I hadn't  thought about... is unfortunate an allocation has to be made per each draw but in the end the user has the control forcing 'with_replacement' or without. I may add a notice though.